kimchi/src/circuits/polynomials/foreign_field_add/circuitgates.rs

```diff
@@ -0,0 +1,243 @@
+    const CONSTRAINTS: u32 = 7;
+
+    fn constraint_checks<T: ExprOps<F>>(env: &ArgumentEnv<F, T>) -> Vec<T> {
+        let field_overflow = env.witness_curr(6);
+        let sign = env.witness_curr(9);
+
+        // Field overflow and sign constraints
+        let mut checks = vec![
+            // Field overflow bit is 0 or s.
+            field_overflow.clone() * (field_overflow - sign.clone()),
+            // Sign flag is 1 or -1
+            (sign.clone() + T::one()) * (sign - T::one()),
+        ];
+
+        // Carry bits are -1, 0, or 1.
+        checks.append(&mut carry(env));
+
+        // Auxiliary inline function to obtain the constraints of a foreign field addition result
+        let mut result = {
+            let foreign_modulus: [T; LIMB_COUNT] = array::from_fn(|i| env.foreign_modulus(i));
+
+            let two_to_limb = T::literal(F::from(TWO_TO_LIMB));
+
+            let left_input_lo = env.witness_curr(0);
+            let left_input_mi = env.witness_curr(1);
+            let left_input_hi = env.witness_curr(2);
+
+            let right_input_lo = env.witness_curr(3);
+            let right_input_mi = env.witness_curr(4);
+            let right_input_hi = env.witness_curr(5);
+
+            let field_overflow = env.witness_curr(6);
+
+            // Result carry bits for limb overflows / underflows.
+            let carry_lo = env.witness_curr(7);
+            let carry_mi = env.witness_curr(8);
+
+            let sign = env.witness_curr(9);
+
+            let result_lo = env.witness_next(0);
+            let result_mi = env.witness_next(1);
+            let result_hi = env.witness_next(2);
+
+            // r_0 = a_0 + s * b_0 - q * f_0 - 2^88 * c_0
+            let result_computed_lo = left_input_lo + sign.clone() * right_input_lo
+                - field_overflow.clone() * foreign_modulus[0].clone()
+                - carry_lo.clone() * two_to_limb.clone();
+            // r_1 = a_1 + s * b_1 - q * f_1 - 2^88 * c_1 + c_0
+            let result_computed_mi = left_input_mi + sign.clone() * right_input_mi
+                - field_overflow.clone() * foreign_modulus[1].clone()
+                - carry_mi.clone() * two_to_limb
+                + carry_lo;
+            // r_2 = a_2 + s * b_2 - q * f_2 + c_1
+            let result_computed_hi = left_input_hi + sign * right_input_hi
+                - field_overflow * foreign_modulus[2].clone()
+                + carry_mi;
+
+            // Result values match
+            vec![
+                result_lo - result_computed_lo,
+                result_mi - result_computed_mi,
+                result_hi - result_computed_hi,
+            ]
+        };
+
+        // Result values match
+        checks.append(&mut result);
+
+        checks
+    }
+}
+
+/// Implementation of the gate for the final bound check for foreign field operations
+/// - Operates on Curr row.
+pub struct ForeignFieldFin<F>(PhantomData<F>);
```

Comment:
This ended up being problematic for a bunch of reasons. 
- The `ForeignFieldAdd` is too general for the type of operation that actually needs to take place just for the bound. This means, it does not force the `sign` to be `1` nor the right input to have that shape. Thus, in order to prove that this is indeed the case we would need to apply some copy constraints to some public input rows, or increase the number of constraints for the bound row specifically. In the latter case, this means the gate type should in any case be a different one. 
- Further modifications in the code needed to take place in order to allow a limb to be 2^88. Meaning, we would need to have right inputs to have 4 limbs as opposed to 3 or some kind of special treatment for limbs having more bits than they should normally have. And in any case, this right input would not work well if the foreign modulus happens to have a specific format where a whole limb is zero. 
- Also, I tried to follow this approach right away but I was finding some issues with the limbs behaviour. Meaning, the checks for the final bound (checking the result of the bound addition) were not satisfied. And the only way I could make the tests pass was by proposing this shorter gate which was kind of a mixture of the old and the ffsub approaches, with the specific constraint for the bound.

Reply:
I don't think we should add a whole extra gate because of how we're writing the tests.

Is it possible to simplify the tests, and create a
```
fn foreign_field_add_gadget<Fp: Field>(x: (Fp, Fp, Fp), y: (Fp, Fp, Fp), sign: Fp, start_row: usize) -> Vec<Gate>
```
that will lay out the witness as needed? This is effectively what we'll end up doing in snarky anyway (albeit using the constraint system so that we don't need to worry about the row). Using that with the final checks should then just be
```
fn foreign_field_add_gadget_bigint(x: BigInt, y: BigInt, start_row: usize) -> Vec<Gate> {
  let x_limbs = limbs_of_bigint(x);
  let y_limbs = limbs_of_bigint(y);
  let mut add_gates = foreign_field_add_gadget(x, y, Fp::zero(), start_row);
  let result_limbs = { /* get limbs from gate witness */ };
  let final_range_check = foreign_field_add_gadget(result_limbs, (0u64.into(), 0u64.into(), 2u64.pow(88).into()));
  /* do permutation stuff */
  add_gates.extend(final_range_check);
  add_gates
}
```
If it's making the tests difficult to write, I'm also in favour of eliding the permutation checks entirely: since we're constructing the witness, we know statically what the permutation is doing without needing the verifier to check it for us.

Reply:
I am not sure I understand the problem... Right now the gadget creator does something similar: it first concatenates foreign field additions for all of the inputs it receives, and then at the very end it extends the gates with the final check for the bound. Except that instead of being a FFAdd gate, it is a FFFin gate. But when writing the tests, you no longer care about the shape of the gadget. You just give the inputs and foreign field, and then the witness is generated accordingly for you.

In any case, the main problem of the approach with the (0,0,2^88) right input was that the constraint checks for the result (meaning that `min_result + 2^264 - modulus = bound`) was not satisfied. I spend a few days (I think Thursday 15 and Monday 19) trying to rewrite the equations but the tests would still fail. And since Jo really wanted this PR to be merged I thought "ok, let's just have a hybrid between both approaches and merge it". I am still intrigued to know why those checks didn't work out, but for FFmul to progress it is really useful that FFadd is at least merged (even if updated to a version with one less gate in the short term). Also this path required less constraints which felt nice. So this is a summary of why the branch has this shape right now 🤔

Reply:
I guess the tricky bit is the `2u64.pow(88).into()` limb.  Would it solve the problem to omit this from range checks?